Mathlib/LinearAlgebra/IsTensorProduct.lean

```diff
@@ -0,0 +1,295 @@
+import Mathlib.LinearAlgebra.TensorProduct
+import Mathlib.GroupTheory.QuotientGroup
+
+set_option autoImplicit false
+
+open TensorProduct MulOpposite
+
+section defs
+
+universe uR uM uN uX
+
+variable (R : Type uR) [Semiring R]
+variable {M : Type uM} [AddCommMonoid M] [Module Rᵐᵒᵖ M]
+variable {N : Type uN} [AddCommMonoid N] [Module R N]
+variable {X : Type uX} [AddCommMonoid X]
+
+variable (M N X) in
+structure BalancedBiAddMonoidHom extends (M →+ N →+ X) where
+(balanced : ∀ (r : R) (m : M) (n : N), toAddMonoidHom (op r • m) n = toAddMonoidHom m (r • n))
+
+instance BalancedBiAddMonoidHom.funLike :
+    FunLike (BalancedBiAddMonoidHom R M N X) M (fun _ => N → X) where
+  coe f := fun m => fun n => f.toAddMonoidHom m n
+  coe_injective' f g h := by
+    cases f
+    cases g
+    congr
+    exact AddMonoidHom.ext fun m => AddMonoidHom.ext fun n => congr_fun (congr_fun h m) n
+
+instance BalancedBiAddMonoidHom.addMonoidHomClass :
+    AddMonoidHomClass (BalancedBiAddMonoidHom R M N X) M (N →+ X) where
+  coe f m := f.toAddMonoidHom m
+  coe_injective' f g h := FunLike.ext _ _ fun m => funext fun n =>
+    FunLike.congr_fun (congr_fun h m) n
+  map_add f _ _ := FunLike.ext _ _ fun _ => by
+    exact FunLike.congr_fun (f.toAddMonoidHom.map_add _ _) _
+  map_zero f := FunLike.ext _ _ fun _ => by
+    exact FunLike.congr_fun (f.toAddMonoidHom.map_zero) _
+
+@[simp] lemma BalancedBiAddMonoidHom.apply_apply_zero (f : BalancedBiAddMonoidHom R M N X) (m : M) :
+    f m 0 = 0 :=
+  (f.toAddMonoidHom m).map_zero
+
+lemma BalancedBiAddMonoidHom.apply_apply_add (f : BalancedBiAddMonoidHom R M N X)
+    (m : M) (n n' : N):
+    f m (n + n') = f m n + f m n' :=
+  (f.toAddMonoidHom m).map_add n n'
+
+@[simp] lemma BalancedBiAddMonoidHom.apply_zero_apply (f : BalancedBiAddMonoidHom R M N X) (n : N) :
+    f 0 n = 0 :=
+  FunLike.congr_fun f.toAddMonoidHom.map_zero _
+
+lemma BalancedBiAddMonoidHom.apply_add_apply (f : BalancedBiAddMonoidHom R M N X)
+    (m m' : M) (n : N):
+    f (m + m') n = f m n + f m' n :=
+  FunLike.congr_fun (f.toAddMonoidHom.map_add m m') n
+
+lemma BalancedBiAddMonoidHom.apply_smul_apply (f : BalancedBiAddMonoidHom R M N X)
+    (r : R) (m : M) (n : N) :
+    f (op r • m) n = f m (r • n) :=
+  f.balanced r m n
+
+variable (tmul : BalancedBiAddMonoidHom R M N X)
+
+class IsTensorProduct :=
+(lift ⦃P : Type uX⦄ [AddCommMonoid P] (_ : BalancedBiAddMonoidHom R M N P) : X →+ P)
+(lift_comp ⦃P : Type uX⦄ [AddCommMonoid P] (f : BalancedBiAddMonoidHom R M N P) (m : M) (n : N) :
+  lift f (tmul m n) = f m n)
+(lift_uniq ⦃P : Type uX⦄ [AddCommMonoid P] (f : BalancedBiAddMonoidHom R M N P) (fHat : X →+ P) :
+  (∀ m n, fHat (tmul m n) = f m n) → lift f = fHat)
```

Comment:
Note we already have [IsTensorProduct](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/IsTensorProduct.html#IsTensorProduct), which is defined in terms of the canonical map from the explicit tensor product being an isomorphism. If you really want to characterize tensor products using a universal property, [here](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Setting.20up.20definitions/near/272082753) are my old thoughts about the universe level restriction.

Reply:
I am just experimenting to see if I can remove  the commutative assumption and write down a characteristic predicate of tensor product over noncommutative rings.

Reply:
I can prove an induction principal using this characteristic predicate
```lean
lemma IsTensorProduct.mem_closure (x : X) :
    x ∈ AddSubgroup.closure { y | ∃ (m : M) (n : N), y = tmul m n } := by

lemma IsTensorProduct.induction {motive : X → Prop}
    (tensor : ∀ m n, motive (tmul m n))
    (add : ∀ x x', motive x → motive x' → motive (x + x'))
    (neg : ∀ x, motive x → motive (-x)) :
    ∀ x, motive x := sorry
```
These depends on `M, N` being `AddCommGroup`, but this is because I didn't figure out how to quotient `AddCommMonoid` by `AddSubmonoid`

Reply:
You'd need to use AddConGen; not every `AddSubmonoid` of `AddCommMonoid ` is the kernel of some AddMonoidHom (unless the monoid is cancellative, I think).

Reply:
The commutativity assumption is easy to remove, but it breaks anyone who wants to take tensor products of algebras until we also fix #7152